django/db/models/functions/comparison.py

```diff
@@ -175,20 +177,24 @@ def join(self, args):
         )
 
     def as_postgresql(self, compiler, connection, **extra_context):
-        if not connection.features.is_postgresql_16:
-            copy = self.copy()
-            copy.set_source_expressions(
+        if (
+            not connection.features.is_postgresql_16
+            or connection.features.uses_server_side_binding
+        ):
+            self.set_source_expressions(
                 [
                     Cast(expression, TextField()) if index % 2 == 0 else expression
-                    for index, expression in enumerate(copy.get_source_expressions())
+                    for index, expression in enumerate(self.get_source_expressions())
                 ]
             )
-            return super(JSONObject, copy).as_sql(
-                compiler,
-                connection,
-                function="JSONB_BUILD_OBJECT",
-                **extra_context,
-            )
+            if not connection.features.is_postgresql_16:
```

Comment:
It's getting complicated :) I'm also not sure why we check `is_postgresql_16` twice :thinking:  We want to use `as_native()` on PostgreSQL 16+ without server-side binding, so maybe it would be clearer to reverse this checks:
```python
if (
    connection.features.is_postgresql_16
    and not connection.features.uses_server_side_binding
):
    return self.as_native(compiler, connection, returning="JSONB", **extra_context)
else:
    copy = self.copy()
    copy.set_source_expressions(
        [
            Cast(expression, TextField()) if index % 2 == 0 else expression
            for index, expression in enumerate(copy.get_source_expressions())
        ]
    )
    return super(JSONObject, copy).as_sql(
        compiler,
        connection,
        function="JSONB_BUILD_OBJECT",
        **extra_context,
    )
```

Reply:
In summary, there are two approaches:
1. have server side bindings use the `as_native` method on PostgreSQL 16+. This means we need to cast the their keys to text and wrap the keys with parenthesis. This PR currently does this
2. have server side bindings use the `JSONB_BUILD_OBJECT` always (your suggestion does this, there is also a way to do this that is essentially a 1 line change)

Given we will backport this is 5.1, and backports should be conservative in order to avoid further regressions, I want to do 2.

I appreciate the work @john-parton has done here around pursuing 1. 
So, I will comment on the ticket with a reference to this approach to keep a record. Folks looking at the git history will be able to find this ticket and that approach if we need to adjust in future

Reply:
Do you want me to open a ticket doing #2 for the 5.2 or a future release?

Reply:
Sure :+1: